Allow cluster to start when only n/2 servers are up

donatello opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

Fixes #3234.

Relaxes the quorum requirement to start the object layer, and skips
quick-healing at start-up (as no write quorum is present).

This fix allows reads to work after start-up, but it does hang on writes. The hang is because no write quorum servers are available and the lock interface currently does not have a possibility of failing/returning an error. In my view, this is a separate issue - #3185 ought to be re-opened.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Toward resolving #3234. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Manually.


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

harshavardhana commented on cmd/prepare-storage.go:
```diff
@@ -150,29 +151,35 @@ func prepForInitXL(firstDisk bool, sErrs []error, diskCount int) InitActions {
 	}
 
-	// Total disks unformatted are in quorum verify if we have some offline disks.
+	// Total disks unformatted are in quorum verify if we have
```
These are not related changes to PR, can we avoid it?

donatello replied:
Done.

harshavardhana commented on cmd/prepare-storage.go:
```diff
@@ -150,29 +151,35 @@ func prepForInitXL(firstDisk bool, sErrs []error, diskCount int) InitActions {
 		}
-		// Some of the formatted disks are possibly corrupted or unformatted, heal them.
+		// Some of the formatted disks are possibly corrupted
+		// or unformatted, heal them.
```
If we still return `WaitForHeal` should the caller proceed to heal? under ReadQuorum?

donatello replied:
From the current implementation in `healFormatXL`, this seems ok as no quorum appears to be required in it - I am not sure if that is entirely correct, but in this case, there seems to be no harm in allowing to heal format.json on a unformatted/corrupted disk if there are readQuorum other formatted disks. E.g.: in a fully working 8-node 8-disk cluster, if 4-disks are replaced, there is no data-loss (due to erasure-coding) and healing should be allowed.

harshavardhana commented on cmd/xl-v1.go:
```diff
@@ -155,6 +155,21 @@ func newXLObjects(storageDisks []StorageAPI) (ObjectLayer, error) {
+	// readQuorum), we cannot perform quick-heal (no
+	// write-quorum). However reads may still be possible, so we
+	// skip quick-heal in this case, and continue.
+	offlineCount := 0
```
perhaps you can use `len(newStorageDisks) - diskCount(newStorageDisks)`  or create a function similar to `diskCount` and rename it as `onlineDiskCount` and this one has `offlineDiskCount`

donatello replied:
Done.

donatello committed:
Allow cluster to start when only n/2 servers are up

Fixes #3234.

Relaxes the quorum requirement to start the object layer, and skips
quick-healing at start-up (as no write quorum is present).

krishnasrinivas commented on cmd/prepare-storage.go:
```diff
@@ -132,13 +132,14 @@ func prepForInitXL(firstDisk bool, sErrs []error, diskCount int) InitActions {
 
 	// No Quorum lots of offline disks, wait for quorum.
-	if disksOffline >= quorum {
+	if disksOffline > readQuorum {
```
Not related to this commit. But when deciding on quorum we have to consider online-disks rather than offline-disks.

Because when data != parity this comparison will be incorrect.

For ex. consider 6-data and 2-parity. readQuorum will be 6. If 3 disks are down, the statement becomes:
```
if 3 > 6 {
    return WaitForQuorum
}
```
Which is incorrect because we should have returned WaitForQuorum.

If we talk in terms of online-disks, it will be like this (which is also more readable):
```
if disksOnline < readQuorum {
    return WaitForQuorum
}
```
which will work as intended.

If/when we make data/parity configurable we will have to fix these things.

harshavardhana replied:
Can this be captured in a bug @krishnasrinivas ?

krishnasrinivas replied:
@harshavardhana opened an umbrella issue https://github.com/minio/minio/issues/4079
